tensornetwork/block_sparse/charge_test.py

```diff
@@ -42,3 +42,238 @@ def test_BaseCharge_copy():
+
+  def run_test():
+    D = 2000
+    B = 6
+    O1 = np.random.randint(-B // 2, B // 2 + 1, D).astype(np.int16)
+    O2 = np.random.randint(-B // 2, B // 2 + 1, D).astype(np.int16)
+    P1 = np.random.randint(0, B + 1, D).astype(np.int16)
+    P2 = np.random.randint(-B // 2, B // 2 + 1, D).astype(np.int16)
+    Q1 = np.random.randint(1, B + 1, D).astype(np.int16)
+    Q2 = np.random.randint(1, B + 1, D).astype(np.int16)
+
+    charges_1 = [O1, O2]
+    charges_2 = [P1, P2]
+    charges_3 = [Q1, Q2]
+
+    fused_1 = fuse_ndarrays(charges_1)
+    fused_2 = fuse_ndarrays(charges_2)
+    fused_3 = fuse_ndarrays(charges_3)
+    q1 = U1Charge(O1) @ U1Charge(P1) @ U1Charge(Q1)
+    q2 = U1Charge(O2) @ U1Charge(P2) @ U1Charge(Q2)
+
+    target = BaseCharge(
+        charges=np.random.randint(-B, B, (3, 1), dtype=np.int16),
+        charge_labels=None,
+        charge_types=[U1Charge, U1Charge, U1Charge])
+    q12 = q1 + q2
+
+    nz_1 = np.nonzero(q12 == target)[0]
+    i1 = fused_1 == target.charges[0, 0]
+    i2 = fused_2 == target.charges[1, 0]
+    i3 = fused_3 == target.charges[2, 0]
+    #pylint: disable=no-member
+    nz_2 = np.nonzero(np.logical_and.reduce([i1, i2, i3]))[0]
+    return nz_1, nz_2
+
+  nz_1, nz_2 = run_test()
+  while len(nz_1) == 0:
+    nz_1, nz_2 = run_test()
+    break
+  assert np.all(nz_1 == nz_2)
+
+
+def test_U1Charge_multiple_fusion():
+
+  def run_test():
+    D = 300
+    B = 4
+    O1 = np.random.randint(-B // 2, B // 2 + 1, D).astype(np.int16)
+    O2 = np.random.randint(-B // 2, B // 2 + 1, D).astype(np.int16)
+    O3 = np.random.randint(-B // 2, B // 2 + 1, D).astype(np.int16)
+    P1 = np.random.randint(0, B + 1, D).astype(np.int16)
+    P2 = np.random.randint(-B // 2, B // 2 + 1, D).astype(np.int16)
+    P3 = np.random.randint(-B // 2, B // 2 + 1, D).astype(np.int16)
+    Q1 = np.random.randint(1, B + 1, D).astype(np.int16)
+    Q2 = np.random.randint(0, B + 1, D).astype(np.int16)
+    Q3 = np.random.randint(-B // 2, B // 2 + 1, D).astype(np.int16)
+
+    charges_1 = [O1, O2, O3]
+    charges_2 = [P1, P2, P3]
+    charges_3 = [Q1, Q2, Q3]
+
+    fused_1 = fuse_ndarrays(charges_1)
+    fused_2 = fuse_ndarrays(charges_2)
+    fused_3 = fuse_ndarrays(charges_3)
+    q1 = U1Charge(O1) @ U1Charge(P1) @ U1Charge(Q1)
+    q2 = U1Charge(O2) @ U1Charge(P2) @ U1Charge(Q2)
+    q3 = U1Charge(O3) @ U1Charge(P3) @ U1Charge(Q3)
+
+    target = BaseCharge(
+        charges=np.random.randint(-B, B, (3, 1), dtype=np.int16),
+        charge_labels=None,
+        charge_types=[U1Charge, U1Charge, U1Charge])
+
+    q123 = q1 + q2 + q3
+
+    nz_1 = np.nonzero(q123 == target)[0]
+    i1 = fused_1 == target.charges[0, 0]
+    i2 = fused_2 == target.charges[1, 0]
+    i3 = fused_3 == target.charges[2, 0]
+    #pylint: disable=no-member
+    nz_2 = np.nonzero(np.logical_and.reduce([i1, i2, i3]))[0]
+    return nz_1, nz_2
+
+  nz_1, nz_2 = run_test()
+  while len(nz_1) == 0:
+    nz_1, nz_2 = run_test()
+  assert np.all(nz_1 == nz_2)
```

Comment:
This test looks like you're doing a lot, but this is the only assert and it seems kind of weak. Are there other checks you can do in the `run_test` function?

Reply:
This tests if charge fusion is correct by fusing charges and comparing the result with a numpy computed fusion. Most of this is initializing, and then recomputing everything directly in numpy to compare with.
nz_1 and nz_2 are large arrays that have to be identical, so I think the test is good. There's not really anything smaller to test here